### ISSUE TITLE
Should List and ListDir return errors?

### ISSUE THREAD

**Issue by ncw** [ncw/rclone#316]
Currently List() returns a channel of Objects.

It is difficult for list to return that an error occurred.

Consider to change the interface so it returns a `chan error` so that the error state can be read after the listing channel has been closed.

**Comment by klauspost** [comment]
Current interface:
```
    // List the Fs into a channel
    List() ObjectsChan

type ObjectsChan chan Object
```

If we are to touch up on this, maybe we should also consider this change, where we send the destination object channel to the list operation:

```
type ObjectsChan struct {
    Objects chan Object
    Errors chan error
    Abort chan struct{}
}

    // List the Fs into a channel
    List(o ObjectsChan) error
```

It will be up to the implementation if they block or not until the Listing has finished. The error on `List` is just intended for immediate errors, for instance if the root folder cannot be found.

There are multiple things that I see as a pros for this approach:
* The caller decide the channel size.
* A caller can send the same channel to multiple operations (though a bit academic at the moment).
* There are more callees than callers, so we (hopefully) reduce code.

I know you spent a lot of time on the current interface, but I just wanted to give my immediate thoughts.

**Comment by ncw** [comment]
Interesting idea....  It is a lot more obvious as an interface without a hidden go routine and with a synchronous error return.

It puts quite a bit more boiler plate into the callers.

```go
objs, errChan := range f.List()
for _, obj := objs {
    // do stuff with obj
}
err := <- errChan
// do stuff with err
```

vs

```go
objs := make(chan Object, 8)
var err error
var wg sync.WaitGroup
wg.Add(1)
go func() {
    defer wg.Done()
    err = f.List(objs)
}()
objs, errChan := range f.List()
for _, obj := objs {
    // do stuff with obj
}
wg.Wait()
// do stuff with err
```

Though the latter code blob would be extremely easy to wrap into a utility function with my original proposed interface, so I think your idea has a lot going for it Klaus.

**Comment by klauspost** [comment]
My thoughts went along the line of a select loop. 

To simplify, lets remote the error as a return value.

```

type ObjectsChan struct {
	Objects chan Object
	Errors  chan error
	Abort   chan struct{}
}

	objs := NewChannelObjects(10) // queue lenght
	go fs.List(objs)

	for {
		select {
		case o, ok := <-objs.Objects:
			if !ok {
				return nil
			}
			// ...
		case err := objs.Errors:
			close(objs.Abort)
			if ok {
				return err
			}
		}
	}
```

It should probably be nested selects, so `obj.Errors` has priority over obj.Objects.

This was the caller doesn't have to create its own goroutine. If we could reduce the boilerplate/error potential for all the fs implementations that would be a great win.

We can augment the "ObjectsChan" with helper functions `SetError(error)` that will help callers synchronize "Errors" and "Abort". We could probably helper functions, so we don't expose any internals, that would make it much easier to avoid racy/deadlocking implementations.

Let me try a concept PR, we just can throw away if we don't like it.

**Comment by ncw** [comment]
Sorry klaus, completely misread your first post!

This would have the advantage that we could add a `Depth` parameter later to `ObjectsChan` also so the remote could limit the depth of the recursion.

`ObjectsChan` probably isn't a good name though, how about `ListOpts` or something like that?

I don't really like there being two ways of returning errors, but I see what you are getting at.

We should try to keep the `List` methods as simple as possible, and be prepared to write some helper methods in `fs`.

A concept PR would be great - we can see how it turns out.

**Comment by klauspost** [comment]
I have added the concept.

From the individual file systems. I found the following:

* The removed need for a separate goroutine in the called function generally make them easier to implement. 
* It is a relatively small conversion. Adding abort functionality is pretty easy as well.
* `fs.Stats.Error()` could also be called automatically from `x.SetError()`, further simplifying error handling.
* Additional parameters will have to be added to the ListOpts interface, it will not be directly accessible fields. Neither pro nor con IMO.

If x.SetError() is called, it will cause x.Add() to return true. There could be a separate `x.Aborted() bool` call, but I didn't see a need for it in the three implementations I did.

Generally, I think it serves what we are looking for. It is very likely some things could probably be tweaked. Let me know what you think.
